internal/resourcecache/resourcecache.go

```diff
@@ -58,14 +62,25 @@ func (rc *ResourceCache) AddResourceToCacheWithTimeout(name string, resource Ide
 	rc.Lock()
 	defer rc.Unlock()
 
-	if _, ok := rc.resources[name]; ok {
-		return errors.Errorf("failed to add entry %s to ResourceCache; entry already exists", name)
+	r, ok := rc.resources[name]
+	// if we don't already have a resource, create it
+	if !ok {
+		r = &Resource{}
+		rc.resources[name] = r
 	}
-	rc.resources[name] = &Resource{
-		resource:     resource,
-		cleanupFuncs: cleanupFuncs,
+	// make sure the resource hasn't already been added to the cache
+	if r.resource != nil || r.cleanupFuncs != nil {
+		return errors.Errorf("failed to add entry %s to ResourceCache; entry already exists", name)
```

Comment:
Shall we really error out if trying to add an entry that already exists to the Cache?
To me it's something we should simply log and sucessfully return.

Reply:
it means that somehow two different goroutines are trying to create the same resource at the same time, and managed to race against each other perfectly. It's a really unlikely scenerio (especially because the locking is perfect :wink: ), but it is definitely an error situation. The most negative thing that would happen is the cleanup funcs for each will be different, so if one overwrote the other, there would be leaked resources